File: scrooge-generator/src/main/resources/scalagen/finagleServiceFunction.scala
private[this] object {{__stats_name}} {
  val RequestsCounter = scopedStats.scope("{{serviceFuncNameForWire}}").counter("requests")
  val SuccessCounter = scopedStats.scope("{{serviceFuncNameForWire}}").counter("success")
  val FailuresCounter = scopedStats.scope("{{serviceFuncNameForWire}}").counter("failures")
  val FailuresScope = scopedStats.scope("{{serviceFuncNameForWire}}").scope("failures")
}
addFunction("{{serviceFuncNameForWire}}", { (iprot: TProtocol, seqid: Int) =>
  try {
    {{__stats_name}}.RequestsCounter.incr()
    val args = {{funcObjectName}}.Args.decode(iprot)
    iprot.readMessageEnd()
    (try {
      iface.{{serviceFuncNameForCompile}}({{argNames}})
    } catch {
      case e: Exception => Future.exception(e)
    }).flatMap { value: {{typeName}} =>
      reply("{{serviceFuncNameForWire}}", seqid, {{funcObjectName}}.Result({{resultNamedArg}}))
    }.rescue {
{{#exceptions}}
      case e: {{exceptionType}} => {
        reply("{{serviceFuncNameForWire}}", seqid, {{funcObjectName}}.Result({{fieldName}} = Some(e)))
      }
{{/exceptions}}
      case e => Future.exception(e)
    }.respond {
      case Return(_) =>
        {{__stats_name}}.SuccessCounter.incr()
      case Throw(ex) =>
        {{__stats_name}}.FailuresCounter.incr()
        {{__stats_name}}.FailuresScope.counter(Throwables.mkString(ex): _*).incr()
    }
  } catch {
    case e: TProtocolException => {
      iprot.readMessageEnd()
      exception("{{serviceFuncNameForWire}}", seqid, TApplicationException.PROTOCOL_ERROR, e.getMessage)
    }
    case e: Exception => Future.exception(e)
  }
})

File: scrooge-generator/src/main/resources/javagen/finagleServiceFunction.java
addFunction("{{serviceFuncNameForWire}}", new Function2<TProtocol, Integer, Future<byte[]>>() {
  public Future<byte[]> apply(TProtocol iprot, final Integer seqid) {
    try {
      {{funcObjectName}}.Args args = {{funcObjectName}}.Args.decode(iprot);
      iprot.readMessageEnd();
      Future<{{typeName}}> result;
      try {
        result = iface.{{serviceFuncNameForCompile}}({{argNames}});
      } catch (Throwable t) {
        result = Future.exception(t);
      }
      return result.flatMap(new Function<{{typeName}}, Future<byte[]>>() {
        public Future<byte[]> apply({{typeName}} value){
          return reply("{{serviceFuncNameForWire}}", seqid, new {{funcObjectName}}.Result.Builder(){{^isVoid}}.success(value){{/isVoid}}.build());
        }
      }).rescue(new Function<Throwable, Future<byte[]>>() {
        public Future<byte[]> apply(Throwable t) {
{{#exceptions}}
          if (t instanceof {{exceptionType}}) {
            return reply("{{ServiceName}}", seqid, new {{funcObjectName}}.Result.Builder().{{fieldName}}(({{exceptionType}}) t).build());
          }
{{/exceptions}}
          return Future.exception(t);
        }
      });
    } catch (TProtocolException e) {
      try {
        iprot.readMessageEnd();
        return exception("{{serviceFuncNameForWire}}", seqid, TApplicationException.PROTOCOL_ERROR, e.getMessage());
      } catch (Exception unrecoverable) {
        return Future.exception(unrecoverable);
      }
    } catch (Throwable t) {
      return Future.exception(t);
    }
  }
});
